---
title: "SKAT: WECARE w/o Danish samples"
author: "AL"
date: "25/01/2021"
output: 
  html_document:
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

# Summary

SKAT for functionaly important variants in bc8 group of genes

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
#options(width=999)
```

# Start section

```{r}

# Time
Sys.time()

# Memory
gc()

# Clean up
rm(list=ls())
graphics.off()

# Options
options(stringsAsFactors = F)

# Working folders
base_folder <- "/Users/alexey/Documents"
project_folder <- file.path(base_folder,"wecare","final_analysis_2021","reanalysis_wo_danish_2021")
scripts_folder <- file.path(project_folder,"scripts","s08_skat")
setwd(scripts_folder)

data_folder <- file.path(project_folder,"data","s08_skat")

# Libraries
library(SKAT)
library(dplyr)

```

# Load data

```{r}

load(file.path(project_folder,"data","s07_pca","s06_update_main_dataset.RData"))

bc8_genes <- scan(file.path(data_folder,"bc8_genes.txt"),what="character")

```


# check_data

```{r check_data}

ls()

dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != phenotypes.df$sample.ID)
sum(rownames(genotypes.mx) != variants.df$ID)

bc8_genes

```

# Select variants 

```{r}

# Variants in the selected genes
bc8 <- variants.df$vep_SYMBOL %in% bc8_genes
sum(bc8) # count all variants in bc8 genes

# Select LoF-s as defined by VEP IMPACT
lof <- variants.df$vep_IMPACT == "HIGH"

# Explore LoF-s in bc8
sum(bc8 & lof) # count LoF-s in bc8
variants.df %>% 
  filter(bc8 & lof) %>% 
  select(ID,vep_SYMBOL,vep_Consequence,vep_IMPACT,CLNSIG,vep_gnomAD_NFE_AF)

# Explore ClinVar annotations in bc8 variants
variants.df %>% 
  filter(bc8) %>% 
  group_by(CLNSIG) %>% 
  summarise(counts=n()) %>% 
  arrange(desc(counts))

# Select known pathogenic variants
likely_pathogenic <- 
  grepl("Pathogenic",variants.df$CLNSIG) | 
  grepl("Likely_pathogenic",variants.df$CLNSIG) |
  grepl("risk_factor",variants.df$CLNSIG)

# Explore known pathogenic variants in bc8
sum(bc8 & likely_pathogenic) # count known pathogenic variants in bc8
variants.df %>% 
  filter(bc8 & likely_pathogenic) %>% 
  select(ID,vep_SYMBOL,vep_Consequence,vep_IMPACT,CLNSIG,vep_gnomAD_NFE_AF)

#bc8_variants.df %>% 
#  group_by(CLNREVSTAT) %>% 
#  summarise(counts=n()) %>% 
#  arrange(desc(counts))

#with_good_support <-
#  variants.df$CLNREVSTAT == "reviewed_by_expert_panel" |
#  variants.df$CLNREVSTAT == "criteria_provided,_multiple_submitters,_no_conflicts" |
#  variants.df$CLNREVSTAT == "criteria_provided,_single_submitter""
#etc
  
# Rare variants
class(variants.df$vep_gnomAD_NFE_AF)
sum(variants.df$vep_gnomAD_NFE_AF==".")
NA -> variants.df[variants.df$vep_gnomAD_NFE_AF==".","vep_gnomAD_NFE_AF"]
variants.df$vep_gnomAD_NFE_AF <- as.numeric(variants.df$vep_gnomAD_NFE_AF)
rare <- variants.df$vep_gnomAD_NFE_AF <= 0.05
sum(rare, na.rm=T)

# Count rare variants in bc8
sum(bc8 & rare, na.rm=T) 

#uncommon <- bc8_variants.df$vep_gnomAD_NFE_AF <= 0.05
#sum(uncommon)

# Known benign
#known_benign <- 
#  grepl("Benign",bc8_variants.df$CLNSIG) | 
#  grepl("Likely_benign",bc8_variants.df$CLNSIG)
#sum(known_benign)

# Functionally important variants
fiv <- bc8 & rare & (lof | likely_pathogenic)
sum(fiv, na.rm=T)

# Clean-up
rm(lof,likely_pathogenic,rare)

```

# Make bc8-fiv dataset

```{r}

bc8_fiv_variants.df <- bc8_variants.df[fiv,] %>% 
  select(ID,vep_SYMBOL,vep_Consequence,vep_IMPACT,CLNSIG,vep_gnomAD_NFE_AF)
bc8_fiv_variants.df
# View(bc8_fiv.df)

bc8_fiv_genotypes


# Clean-up
rm(lof,likely_pathogenic,rare,fiv)

```

# prepare_data_for_analysis

No missed data in outcomes, covariates or eigenvectors.  

Missed genotypes are imputed by SKAT as average AF  
(default impute.method="fixed" uses mean genotype values).  

```{r prepare_data_for_analysis}

# Outcomes
Y <- phenotypes.df$cc
sum(is.na(Y)) # 0

# Genotypes: transpose non-aggregated variants
G <- t(genotypes.mx[,])

# Do NOT impute missed genotypes (they are handled by SKAT)
sum(is.na(G))/(ncol(G)*nrow(G)) # ~2.3%

# Eigenvectors
E <- as.matrix(phenotypes.df[,c("eig1","eig2")])
sum(is.na(E)) # 0

# "Matching" covariates 

# Originally, the cases and controls were matched by Age at diagnosis and 
# time to event / lack of event ("age_dx", "rstime"). 
# Hence the "M" abbreviation (called SKAT-M in preliminary analyses). 

# At the same time, SKAT does not allow using this matching information.  
# Also, some samples failed sequencing, thus breaking the pairs. 
# Overall, the compared groups are very well balanced by "age_dx", "rstime". 
# Still, we desided to use these covariates to account for differences at 
# individual level.  

# By including these we assume linear effect of these confaunders.  
# This is fully supported by the previous biological knowledge. 

M <- as.matrix(phenotypes.df[,c("age_dx", "rstime")])
sum(is.na(M)) # 0

# Treatment: ether endo or chemo or both 
# There is one case with missed information on hormonal treatment 
# However, it has chemo, so the overall trearment has no missed data 
hormone <- phenotypes.df$hormone
sum(is.na(hormone))
chemo <- phenotypes.df$chemo_cat != "no"
sum(is.na(chemo))
treatment <- as.integer(chemo | hormone)
sum(is.na(treatment)) # 0

# Number of pregnancies
num_preg <- as.integer(phenotypes.df$num_preg)
sum(is.na(num_preg)) # 0

# Lobular pathology - excluded because of a missed value?
# Also: has a known genetic association (CDH1)
#lobular <- phenotypes.df$hist_cat == "lobular"
#sum(is.na(lobular)) # 1

# Combined "Clinical" covariates
C <- cbind(treatment, num_preg)
sum(is.na(C)) # 0

# List of genes 
genes <- unique(as.character(variants.df$SYMBOL))
length(genes)
sum(is.na(genes))

# Clean-up
rm(chemo, hormone, treatment, num_preg) # lobular

```

# prepare_empty_matrix_for_results

```{r prepare_empty_matrix_for_results}

# Prepare header
header=c("gene", "num_var", 
         "svt_p", "svt_is_accurate", "svt_map", 
         "burden_p",  "burden_is_accurate", "burden_map", 
         "skat_p", "skat_is_accurate", "skat_map")

# Write header to output matrix 
# (do not use data frame here: it may cause issues with rbind)
result.mx <- matrix(ncol=11, nrow=0) 
colnames(result.mx) <- header

# Clean-up
rm(header)

```

# SKAT

Assuming that no errors generated by the SKAT functions (accertained in a separate script)  
Ignoring warnings about imputation and flipping (captured and assessed in a separate script)  

Takes ~30 min  

```{r SKAT, warning=FALSE}

# Calculate SKAT null model (the same for all genes and all tests)
skat_null <- SKAT_Null_Model(Y ~ C + E + M, out_type="D")
# D for the dichotomous outcome

# For each gene
for(gene in genes){

  # Get variants
  vars <- as.vector(variants.df[variants.df$SYMBOL==gene, "SplitVarID"])
  
  # Get number of variants
  num_vars <- length(vars)
    
  # Get vector of genotypes  
  X <- G[,vars]

  # If there is one variant only
  if(num_vars == 1){

    # Aggregating variants (burden or skat) is irrelevant
    burden_p <- NA
    burden_is_accurate <- NA
    burden_map <- NA
    skat_p <- NA
    skat_is_accurate <- NA
    skat_map <- NA

    # Calculate Single-variant test (svt)
    svt <- SKATBinary_Single(X, skat_null, method.bin = "ER")
    svt_p <- svt$p.value
    svt_is_accurate <- svt$is.accurate # Was resampling suficient ?
    svt_map <- svt$MAP # Minimal Acheaved P-value (for adjusted QQ-plot)
    rm(svt)
    
  # If there are several variants in the gene  
  }else{
    
    # Single-variant test is irrelevant
    svt_p <- NA
    svt_is_accurate <- NA
    svt_map <- NA

    # Calculate burden test with default weights beta[1,25]
    burden_test <- SKATBinary(X, skat_null, method="Burden", method.bin = "ER")
    burden_p <- burden_test$p.value
    burden_is_accurate <- burden_test$is.accurate # Was resampling suficient ?
    burden_map <- burden_test$MAP # Minimal Acheaved P-value (for adjusted QQ-plot)
    rm(burden_test)

    # Calculate SKAT variance-based test with default weights beta[1,25]
    skat_test <- SKATBinary(X, skat_null, method="SKAT", method.bin = "ER")
    skat_p <- skat_test$p.value
    skat_is_accurate <- skat_test$is.accurate # Was resampling suficient ?
    skat_map <- skat_test$MAP # Minimal Acheaved P-value (for adjusted QQ-plot)
    rm(skat_test)
    
  }
  
  # Add results to matrix
  result.mx <- rbind(result.mx, c(gene, num_vars, 
                svt_p, svt_is_accurate, svt_map, 
                burden_p, burden_is_accurate, burden_map, 
                skat_p, skat_is_accurate, skat_map))

  # Clean-up
  rm(vars, num_vars, X, 
     svt_p, svt_is_accurate, svt_map, 
     burden_p, burden_is_accurate, burden_map, 
     skat_p, skat_is_accurate, skat_map)

} # next gene

# explore result
dim(result.mx)

# Clean-up
rm(Y, G, E, C, M, gene, genes, skat_null)

```

# explore_update_and_save_result

```{r explore_update_and_save_result}

# --- Convert to dataframe, change factors to vectors --- #

genes_aggr_skat.df <- as.data.frame(result.mx)
rownames(genes_aggr_skat.df) <- genes_aggr_skat.df$gene

genes_aggr_skat.df$gene <- as.vector(genes_aggr_skat.df$gene)
genes_aggr_skat.df$num_var <- as.integer(as.vector(genes_aggr_skat.df$num_var))

genes_aggr_skat.df$svt_p <- as.double(as.vector(genes_aggr_skat.df$svt_p))
genes_aggr_skat.df$svt_is_accurate <- as.logical(as.vector(genes_aggr_skat.df$svt_is_accurate))
genes_aggr_skat.df$svt_map <- as.double(as.vector(genes_aggr_skat.df$svt_map))

genes_aggr_skat.df$burden_p <- as.double(as.vector(genes_aggr_skat.df$burden_p))
genes_aggr_skat.df$burden_is_accurate <- as.logical(as.vector(genes_aggr_skat.df$burden_is_accurate))
genes_aggr_skat.df$burden_map <- as.double(as.vector(genes_aggr_skat.df$burden_map))

genes_aggr_skat.df$skat_p <- as.double(as.vector(genes_aggr_skat.df$skat_p))
genes_aggr_skat.df$skat_is_accurate <- as.logical(as.vector(genes_aggr_skat.df$skat_is_accurate))
genes_aggr_skat.df$skat_map <- as.double(as.vector(genes_aggr_skat.df$skat_map))

# Check data
dim(genes_aggr_skat.df)
str(genes_aggr_skat.df)
sum(is.null(genes_aggr_skat.df)) # no null data (just in case...)
summary(genes_aggr_skat.df) # all tests are accurate

# Exclude "is_accurate" columns (because all tests are accurate - see above)
colnames(genes_aggr_skat.df)
genes_aggr_skat.df <- genes_aggr_skat.df[,c(-4,-7,-10)]
colnames(genes_aggr_skat.df)

# --- Combine svt with burden/skat --- #

# Compile the combined data 
svt_burden_p <- ifelse(is.na(genes_aggr_skat.df$svt_p), genes_aggr_skat.df$burden_p, genes_aggr_skat.df$svt_p)
sum(is.na(svt_burden_p))
length(svt_burden_p)

svt_burden_map <- ifelse(is.na(genes_aggr_skat.df$svt_map), genes_aggr_skat.df$burden_map, genes_aggr_skat.df$svt_map)
sum(is.na(svt_burden_map))
length(svt_burden_map)

svt_skat_p <- ifelse(is.na(genes_aggr_skat.df$svt_p), genes_aggr_skat.df$skat_p, genes_aggr_skat.df$svt_p)
sum(is.na(svt_skat_p))
length(svt_skat_p)

svt_skat_map <- ifelse(is.na(genes_aggr_skat.df$svt_map), genes_aggr_skat.df$skat_map, genes_aggr_skat.df$svt_map)
sum(is.na(svt_skat_map))
length(svt_skat_map)

# Add combined data to results table
genes_aggr_skat.df <- cbind(genes_aggr_skat.df, svt_burden_p, svt_burden_map, svt_skat_p, svt_skat_map)
colnames(genes_aggr_skat.df)

# --- Write results to text file --- #

results_file <- paste(base_folder, "results", "r12_skat_wecare_only.txt", sep="/")
write.table(genes_aggr_skat.df, file=results_file, quote=FALSE, sep="\t")

# --- Clean-up --- #

rm(result.mx, results_file, svt_burden_p, svt_burden_map, svt_skat_p, svt_skat_map)

```

# compare_p_estimates_between_methods

```{r compare_p_estimates_between_methods}

# svt vs burden vs skat
mean_p_values <- 
  c(mean(genes_aggr_skat.df$svt_p, na.rm=TRUE), 
    mean(genes_aggr_skat.df$burden_p, na.rm=TRUE),
    mean(genes_aggr_skat.df$skat_p, na.rm=TRUE))
names(mean_p_values) <- c("single", "burden", "skat")
barplot(mean_p_values, main="Compare mean p-values, SKAT M")

# burden vs skat
plot(genes_aggr_skat.df$burden_p, 
     genes_aggr_skat.df$skat_p, 
     main="P-values: Burden vs SKAT",
     xlab="Burden", ylab="SKAT")

cor.test(genes_aggr_skat.df$burden_p, 
         genes_aggr_skat.df$skat_p)

# Clean-up
rm(mean_p_values)

```

# qq_plots

Remove NAs because they may interfere with the QQ plots  

```{r qq_plots}

# ----- svt only ----- #

# Remove NAs
svt_p <- genes_aggr_skat.df$svt_p[!is.na(genes_aggr_skat.df$svt_p)]
length(svt_p) # 4,619
summary(svt_p)

svt_map <- genes_aggr_skat.df$svt_map[!is.na(genes_aggr_skat.df$svt_map)]
length(svt_map) # 4,619
summary(svt_map)

# QQ plot with unadjusted
QQPlot_Adj(svt_p, svt_map, 
           main=paste("CBC vs UBC: Single variant test only\n", 
                      format(length(svt_p), big.mark=","), 
                      "genes with 1 variant"))

# QQ plot without unadjusted
QQPlot_Adj(svt_p, svt_map, Is.unadjsted=FALSE, 
           main=paste("CBC vs UBC: Single variant test only\n", 
                      format(length(svt_p), big.mark=","), 
                      "genes with 1 variant"))

# ----- burden only ----- #

# Remove NAs
burden_p <- genes_aggr_skat.df$burden_p[!is.na(genes_aggr_skat.df$burden_p)]
length(burden_p) # 4,030
summary(burden_p)

burden_map <- genes_aggr_skat.df$burden_map[!is.na(genes_aggr_skat.df$burden_map)]
length(burden_map) # 4,030
summary(burden_map)

# QQ plot with unadjusted
QQPlot_Adj(burden_p, burden_map, 
           main=paste("CBC vs UBC: Burden test only\n", 
                      format(length(burden_p), big.mark=","), 
                      "genes with >1 variant"))

# QQ plot without unadjusted
QQPlot_Adj(burden_p, burden_map, Is.unadjsted=FALSE, 
           main=paste("CBC vs UBC: Burden test only\n", 
                      format(length(burden_p), big.mark=","), 
                      "genes with >1 variant"))

# ----- skat only ----- #

# Remove NAs
skat_p <- genes_aggr_skat.df$skat_p[!is.na(genes_aggr_skat.df$skat_p)]
length(skat_p) # 4,030
summary(skat_p)

skat_map <- genes_aggr_skat.df$skat_map[!is.na(genes_aggr_skat.df$skat_map)]
length(skat_map) # 4,030
summary(skat_map)

# QQ plot with unadjusted
QQPlot_Adj(skat_p, skat_map, 
           main=paste("CBC vs UBC: Skat test only\n", 
                      format(length(skat_p), big.mark=","), 
                      "genes with >1 variant"))

# QQ plot without unadjusted
QQPlot_Adj(skat_p, skat_map, Is.unadjsted=FALSE, 
           main=paste("CBC vs UBC: Skat test only\n", 
                      format(length(skat_p), big.mark=","), 
                      "genes with >1 variant"))

# ----- combined svt-burden ----- #

# Check absence of NAs
svt_burden_p <- genes_aggr_skat.df$svt_burden_p
sum(is.na(svt_burden_p)) # 0
length(svt_burden_p) # 8,649
summary(svt_burden_p)

svt_burden_map <- genes_aggr_skat.df$svt_burden_map
sum(is.na(svt_burden_map)) # 0
length(svt_burden_map) # 8,649
summary(svt_burden_map)

# QQ plot with unadjusted
QQPlot_Adj(svt_burden_p, svt_burden_map, 
           main=paste("CBC vs UBC: combined SVT and Burden tests\nAll", 
                      format(length(svt_burden_p), big.mark=","), "genes"))

# QQ plot without unadjusted
QQPlot_Adj(svt_burden_p, svt_burden_map, Is.unadjsted=FALSE, 
           main=paste("CBC vs UBC: combined SVT and Burden tests\nAll", 
                      format(length(svt_burden_p), big.mark=","), "genes"))

# ----- combined svt-skat ----- #

# Check absence of NAs
svt_skat_p <- genes_aggr_skat.df$svt_skat_p
sum(is.na(svt_skat_p)) # 0
length(svt_skat_p) # 8,649
summary(svt_skat_p)

svt_skat_map <- genes_aggr_skat.df$svt_skat_map
sum(is.na(svt_skat_map)) # 0
length(svt_skat_map) # 8,649
summary(svt_skat_map)

# QQ plot with unadjusted
QQPlot_Adj(svt_skat_p, svt_skat_map, 
           main=paste("CBC vs UBC: combined SVT and Skat tests\nAll", 
                      format(length(svt_skat_p), big.mark=","), "genes"))

# QQ plot without unadjusted
QQPlot_Adj(svt_skat_p, svt_skat_map, Is.unadjsted=FALSE, 
           main=paste("CBC vs UBC: combined SVT and Skat tests\nAll", 
                      format(length(svt_skat_p), big.mark=","), "genes"))

# --- Clean-up --- #

rm(svt_p, svt_map, burden_p, burden_map, skat_p, skat_map, svt_burden_p, svt_burden_map, svt_skat_p, svt_skat_map)

```

# data_summary

```{r data_summary}

ls()

dim(genotypes.mx)
class(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(kgen.df)
colnames(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
colnames(exac.df)
exac.df[1:5,1:5]

dim(variants.df)
str(variants.df)
variants.df[1:5,1:5]

dim(phenotypes.df)
str(phenotypes.df)
phenotypes.df[1:5,1:5]

dim(genes_aggr_skat.df)
str(genes_aggr_skat.df)
genes_aggr_skat.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != rownames(phenotypes.df))

sum(rownames(genotypes.mx) != rownames(kgen.df))
sum(rownames(genotypes.mx) != rownames(exac.df))
sum(rownames(genotypes.mx) != rownames(variants.df))

```

# save_results

```{r save_results}

save.image(paste(base_folder, "results", "r12_skat_wecare_only.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
