---
title: "WES: BC8 crude counts"
author: "Alexey Larionov"
date: "27 January 2021"
output:
  html_document:
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

- Select pthogenic variants in BC8
- Calculate crude counts: totals, and separately for cases and controls  
- Fisher test for the crude counts  

Check call rates (per variant and per sample) before and after filtering.  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
#options(width=999)
```

# Start section

```{r}

# Time
Sys.time()

# Memory
gc()

# Clean up
rm(list=ls())
graphics.off()

# Options
options(stringsAsFactors = F)

# Working folders
base_folder <- "/Users/alexey/Documents"
project_folder <- file.path(base_folder,"wecare","final_analysis_2021","reanalysis_wo_danish_2021")

scripts_folder <- file.path(project_folder,"scripts","s08_crude_counts")
setwd(scripts_folder)

data_folder <- file.path(project_folder,"data","s08_crude_counts")

# Libraries
library(dplyr)
library(ggVennDiagram)

```

# Load data

```{r}

load(file.path(data_folder,"s01_crude_counts_per_variant.RData"))

# Update working folders
base_folder <- "/Users/alexey/Documents"
project_folder <- file.path(base_folder,"wecare","final_analysis_2021","reanalysis_wo_danish_2021")
scripts_folder <- file.path(project_folder,"scripts","s08_crude_counts")
data_folder <- file.path(project_folder,"data","s08_crude_counts")

bc8_genes <- scan(file.path(data_folder,"bc8_genes.txt"),what="character")

```

# Check data

```{r}

ls()

# Check objects dimentions
dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Check that the tables and matrices are in sync
sum(phenotypes.df$wes_id != colnames(genotypes.mx))
sum(variants.df$ID != rownames(genotypes.mx))

# Check list of genes
bc8_genes

```

# Select variants 

## Keep only variants in the selected genes

```{r}

bc8_variants.df <- variants.df[variants.df$vep_SYMBOL %in% bc8_genes,]
dim(bc8_variants.df)

```

## LoF-s as defined by VEP IMPACT

```{r}

lof <- bc8_variants.df$vep_IMPACT == "HIGH"
sum(lof)

```

## Select known (likely) pathogenic by ClinVar

```{r}

# Explore ClinVar annotations
bc8_variants.df %>% 
  group_by(CLNSIG) %>% 
  summarise(counts=n()) %>% 
  arrange(desc(counts))

# Select pathogenic variants
pathogenic <- 
  grepl("Pathogenic",bc8_variants.df$CLNSIG) | 
  grepl("Likely_pathogenic",bc8_variants.df$CLNSIG)
sum(pathogenic)

#grepl("risk_factor",bc8_variants.df$CLNSIG)

#bc8_variants.df %>% 
#  group_by(CLNREVSTAT) %>% 
#  summarise(counts=n()) %>% 
#  arrange(desc(counts))

#with_good_support <-
#  variants.df$CLNREVSTAT == "reviewed_by_expert_panel" |
#  variants.df$CLNREVSTAT == "criteria_provided,_multiple_submitters,_no_conflicts" |
#  variants.df$CLNREVSTAT == "criteria_provided,_single_submitter""
#etc

```

## Rare variants

```{r}

# Rare in gnomeD NFE
sum(bc8_variants.df$vep_gnomAD_NFE_AF <= 0.01, na.rm=T)
sum(is.na(bc8_variants.df$vep_gnomAD_NFE_AF))

rare_in_gnomAD_NFE <- bc8_variants.df$vep_gnomAD_NFE_AF <= 0.01 | 
  is.na(bc8_variants.df$vep_gnomAD_NFE_AF)
sum(rare_in_gnomAD_NFE)

# Rare in dataset
rare_in_wecare <- bc8_variants.df$total_af <= 0.01
sum(rare_in_wecare)

# Rare in controls
#rare_in_controls <- bc8_variants.df$controls_ac <= 0.05
#sum(rare_in_controls)

sets.ls <- list(gnomAD_NFE=bc8_variants.df$ID[rare_in_gnomAD_NFE],
            wecare=bc8_variants.df$ID[rare_in_wecare])

ggVennDiagram(sets.ls)

#rare <- rare_in_gnomAD_NFE | rare_in_wecare
#rare <- rare_in_gnomAD_NFE & rare_in_wecare
rare <- rare_in_gnomAD_NFE
#rare <- rare_in_controls

#uncommon <- bc8_variants.df$vep_gnomAD_NFE_AF <= 0.05
#sum(uncommon)

rm(rare_in_gnomAD_NFE, rare_in_wecare, sets.ls)

```

## Known benign or conflicting_interpretations

known_benign not needed because of conflicting would include them

```{r}

# Known benign
#known_benign <- 
#  grepl("Benign",bc8_variants.df$CLNSIG) | 
#  grepl("Likely_benign",bc8_variants.df$CLNSIG)
#sum(known_benign)

conflicting_interpretations <- 
  grepl("Conflicting_interpretations",bc8_variants.df$CLNSIG)

sum(conflicting_interpretations)

```

## Combined functionaly important

```{r}
# Functionally important variants
fiv <- (lof | pathogenic) & rare & !conflicting_interpretations
sum(fiv)

bc8_fiv.df <- bc8_variants.df[fiv,] %>% 
  select(ID,vep_SYMBOL,vep_Consequence,vep_IMPACT,CLNSIG,CLNREVSTAT,
         vep_EXON,vep_gnomAD_NFE_AF,
         total_ac,total_an,total_af,
         controls_ac,controls_an,controls_af,
         cases_ac,cases_an,cases_af)
bc8_fiv.df
# View(bc8_fiv.df)

# Clean-up
rm(lof,pathogenic,rare,fiv,conflicting_interpretations)

```

# Aggregated crude counts and Fisher test

```{r}

# Calculate crude counts
controls_ac <- sum(bc8_fiv.df$controls_ac) 
controls_an <- sum(bc8_fiv.df$controls_an)
controls_af <- controls_ac / controls_an

cases_ac <- sum(bc8_fiv.df$cases_ac)
cases_an <- sum(bc8_fiv.df$cases_an)
cases_af <- cases_ac / cases_an

# Calculate Fisher test
fisher_matrix <- matrix(c(cases_ac, cases_an - cases_ac,
                          controls_ac, controls_an - controls_ac),
                          nrow=2, byrow=T)

colnames(fisher_matrix) <- c("Alt","Ref")
rownames(fisher_matrix) <- c("Cases","Controls")
fisher_matrix 

ft <- fisher.test(fisher_matrix, conf.int=T)

ft

afs <- c(controls_af,cases_af)
names(afs) <- c("Controls","Cases")

cc <- table(phenotypes.df$cc)

mp <- barplot(afs, ylim=c(0,max(afs)+0.0005), 
main=paste("WECARE WES w/o Danish\n",cc[1],"controls vs ",cc[2]," cases\nCrude counts Fisher p=",round(ft$p.value,3)), ylab="Mean AF in selected variants")

text(mp,afs+0.0003,round(afs,4))

# Clean-up
rm(controls_ac, controls_an, controls_af,
   cases_ac, cases_an, cases_af,
   afs, mp, fisher_matrix,cc,ft)

```

# Carriers of the selected variants

```{r}

genotypes_bc8.mx <- genotypes.mx[bc8_fiv.df$ID,]
dim(genotypes_bc8.mx)

carriers.udf <- function(x){paste(names(x)[which(x>0)], collapse=",")}
x <- apply(genotypes_bc8.mx,1,carriers.udf)
x

rm(x)

```

# Check data

```{r}

ls()

# Check objects dimentions
dim(bc8_variants.df)
dim(bc8_fiv.df)

dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Check that the tables and matrices are in sync
sum(phenotypes.df$wes_id != colnames(genotypes.mx))
sum(variants.df$ID != rownames(genotypes.mx))

```

# Save results

```{r}

save.image(file.path(data_folder,"s02_crude_counts_bc8.RData"))

```

# Final section

```{r}

sessionInfo()
Sys.time()
gc()

```
